src/encoded/search.py

```diff
@@ -280,25 +283,70 @@ def execute_search_for_all_results(search):
     return es_result
 
 
-def normalize_query(request, search_type):
+def normalize_query(request, types, doc_types):
     """
-    Normalize the query used to make the search. If no type is provided,
-    use type=Item
+    Normalize the query by calculating and setting request.normalized_params
+    (a webob MultiDict) that is derived from custom query rules and also
+    the list of doc_types specified by set_doc_types(). The normalize_param
+    helper function finds field_schema for each query parameter and enforces
+    a set of rules (see below). If the query item types differ from doc_types,
+    override with doc_types
+
+    Args:
+        request: the current Request
+        types: registry[TYPES]
+        doc_types (list): item_types to use for the search
+
+    Returns:
+        string: query string built from normalized params
     """
-    item_type = search_type if search_type != None else 'Item'
-    types = request.registry[TYPES]
-    fixed_types = (
-        (k, types[v].name if k == 'type' and v in types else v)
+    def normalize_param(key, val):
+        """
+        Process each key/val in the original query param. As part of this,
+        obtain the field schema for each parameter.
+        Current rules:
+        - for 'type', get name from types (from the registry)
+        - append '.display_title' to any terminal linkTo query field
+        """
+        field_schema = schema_for_field(key, request, doc_types)
+        # must drill down into arrays/subobjects for the given field
+        while field_schema and ('items' in field_schema or 'properties' in field_schema):
+            try:
+                field_schema = field_schema['items']
+            except KeyError:
+                pass
+            try:
+                field_schema = field_schema['properties']
+            except KeyError:
+                pass
+        if key == 'type' and val in types:
```

Comment:
Maybe this could be at top of function, before `field_schema = ...` is calculated?
Since neither the if check nor its return val require the field_schema to yet be calculated, and `type=` is a URI param which would almost always be present, it might make sense performance-wise.

Reply:
Sure thing. I think the performance increase is negligible but would probably be cleaner